File: src/main/groovy/org/fbgroups/controllers/util/SocialControllerUtil.java
package org.fbgroups.controllers.util;

import org.fbgroups.FBUtils;
import org.fbgroups.entity.UserConnection;
import org.fbgroups.entity.UserConnectionRepository;
import org.fbgroups.entity.UserProfile;
import org.fbgroups.entity.UserProfileRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.web.WebAttributes;
import org.springframework.social.facebook.api.Facebook;
import org.springframework.social.facebook.api.Group;
import org.springframework.social.facebook.api.PagedList;
import org.springframework.social.facebook.api.Post;
import org.springframework.stereotype.Component;
import org.springframework.ui.Model;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import java.security.Principal;
import java.util.Enumeration;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Created by magnus on 2014-08-24.
 */
@Component
public class SocialControllerUtil {

    private static final Logger LOG = LoggerFactory.getLogger(SocialControllerUtil.class);

    private static final String USER_CONNECTION = "MY_USER_CONNECTION";
    private static final String USER_PROFILE = "MY_USER_PROFILE";

    @Autowired
    UserConnectionRepository userConnectionRepository;

    @Autowired
    org.springframework.social.connect.UsersConnectionRepository springConnectionRepo;

    @Autowired
    UserProfileRepository userProfileRepository;


    @Autowired
    private UsersDao usersDao;

    public void setModel(HttpServletRequest request, Principal currentUser, Model model) {

        // SecurityContext ctx = (SecurityContext) session.getAttribute("SPRING_SECURITY_CONTEXT");

        String userId = currentUser == null ? null : currentUser.getName();
        String path = request.getRequestURI();
        HttpSession session = request.getSession();

        UserConnection connection = null;
        UserProfile profile = null;
        String displayName = null;
        String data = null;

        // Collect info if the user is logged in, i.e. userId is set
        if (userId != null) {

            // Get the current UserConnection from the http session
            connection = getUserConnection(session, userId);

            // Get the current UserProfile from the http session
            profile = getUserProfile(session, userId);

            // Compile the best display name from the connection and the profile
            displayName = getDisplayName(connection, profile);

        }

        Throwable exception = (Throwable) session.getAttribute(WebAttributes.AUTHENTICATION_EXCEPTION);

        // Update the model with the information we collected
        model.addAttribute("exception", exception == null ? null : exception.getMessage());
        model.addAttribute("currentUserId", userId);
        model.addAttribute("currentUserProfile", profile);
        model.addAttribute("currentUserConnection", connection);
        model.addAttribute("currentUserDisplayName", displayName);

        if (LOG.isDebugEnabled()) {
            logInfo(request, model, userId, path, session);
        }
    }

    protected void logInfo(HttpServletRequest request, Model model, String userId, String path, HttpSession session) {
        // Log the content of the model
        LOG.debug("Path: " + path + ", currentUserId: " + userId);

        LOG.debug("Non-null request-attributes:");
        for (Enumeration<String> rane = request.getAttributeNames(); rane.hasMoreElements(); ) {
            String key = rane.nextElement();
            Object value = session.getAttribute(key);
            if (value != null) {
                LOG.debug(" - " + key + " = " + value);
            }
        }

        LOG.debug("Session-attributes:");
        for (Enumeration<String> sane = session.getAttributeNames(); sane.hasMoreElements(); ) {
            String key = sane.nextElement();
            LOG.debug(" - " + key + " = " + session.getAttribute(key));
        }

        Set<Map.Entry<String, Object>> me = model.asMap().entrySet();
        LOG.debug("ModelElements (" + me.size() + "):");
        for (Map.Entry<String, Object> e : me) {
            LOG.debug(" - " + e.getKey() + " = " + e.getValue());
        }
    }

    /**
     * Get the current UserProfile from the http session
     *
     * @param session
     * @param userId
     * @return
     */
    protected UserProfile getUserProfile(HttpSession session, String userId) {
        UserProfile profile = (UserProfile) session.getAttribute(USER_PROFILE);

        // Reload from persistence storage if not set or invalid (i.e. no valid userId)
        if (profile == null || !userId.equals(profile.getUserid())) {
            profile = userProfileRepository.getOne(userId);
            session.setAttribute(USER_PROFILE, profile);
        }
        return profile;
    }

    /**
     * Get the current UserConnection from the http session
     *
     * @param session
     * @param userId
     * @return
     */
    UserConnection getUserConnection(HttpSession session, String userId) {
        UserConnection connection;
        connection = (UserConnection) session.getAttribute(USER_CONNECTION);

        // Reload from persistence storage if not set or invalid (i.e. no valid userId)
        if (connection == null || !userId.equals(connection.getUserid())) {
            connection = userConnectionRepository.getOne(userId);
            session.setAttribute(USER_CONNECTION, connection);
        }
        return connection;
    }

    /**
     * Compile the best display name from the connection and the profile
     *
     * @param connection
     * @param profile
     * @return
     */
    String getDisplayName(UserConnection connection, UserProfile profile) {

        // The name is set differently in different providers so we better look in both places...
        if (connection.getDisplayname() != null) {
            return connection.getDisplayname();
        } else {
            return profile.getName();
        }
    }

    public void printGroups(Principal currentUser, HttpServletRequest request) {
        String userId = currentUser == null ? null : currentUser.getName();
        // Get the current UserConnection from the http session
        UserConnection connection = getUserConnection(request.getSession(), userId);
        Facebook api = FBUtils.getApi(springConnectionRepo, userId);
        List<Group> groupList = api.fetchConnections("me", "groups", Group.class);
        groupList.forEach(g->{
            System.out.println(g.getName());
            PagedList<Post> feed = api.feedOperations().getFeed(g.getId());
            feed.forEach(p-> System.out.println("\t" + p.getMessage()));
        });

    }

}

File: src/main/kotlin/org/fbgroups/controllers/util/SocialControllerUtil.kt
package org.fbgroups.controllers.util

import org.fbgroups.entity.UserConnection
import org.fbgroups.entity.UserConnectionRepository
import org.fbgroups.entity.UserProfile
import org.fbgroups.entity.UserProfileRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.web.WebAttributes
import org.springframework.stereotype.Component
import org.springframework.ui.Model
import java.security.Principal
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpSession

/**
 * Created by magnus on 2014-08-24.
 */
@Component
class SocialControllerUtil {

    @Autowired
    lateinit var userConnectionRepository: UserConnectionRepository

    @Autowired
    lateinit var userProfileRepository: UserProfileRepository

    fun setModel(request: HttpServletRequest, currentUser: Principal?, model: Model) {

        // SecurityContext ctx = (SecurityContext) session.getAttribute("SPRING_SECURITY_CONTEXT");

        val userId = currentUser?.name
        val path = request.requestURI
        val session = request.session

        var connection: UserConnection? = null
        var profile: UserProfile? = null
        var displayName: String? = null

        // Collect info if the user is logged in, i.e. userId is set
        if (userId != null) {

            // Get the current UserConnection from the http session
            connection = getUserConnection(session, userId)

            // Get the current UserProfile from the http session
            profile = getUserProfile(session, userId)

            // Compile the best display name from the connection and the profile
            displayName = getDisplayName(connection, profile)

        }

        val exception = session.getAttribute(WebAttributes.AUTHENTICATION_EXCEPTION) as Throwable?

        // Update the model with the information we collected
        model.addAttribute("exception", exception?.message)
        model.addAttribute("currentUserId", userId)
        model.addAttribute("currentUserProfile", profile)
        model.addAttribute("currentUserConnection", connection)
        model.addAttribute("currentUserDisplayName", displayName)
    }

    /**
     * Get the current UserProfile from the http session
     *
     * @param session
     * @param userId
     * @return
     */
    protected fun getUserProfile(session: HttpSession, userId: String): UserProfile {
        val profile: UserProfile = session.getAttribute(USER_PROFILE) as UserProfile?
                ?: userProfileRepository.getOne(userId)

        session.setAttribute(USER_PROFILE, profile)

        return profile
    }

    /**
     * Get the current UserConnection from the http session
     *
     * @param session
     * @param userId
     * @return
     */
    internal fun getUserConnection(session: HttpSession, userId: String?): UserConnection {
        val connection: UserConnection =
                (session.getAttribute(USER_CONNECTION) as UserConnection?) ?: userConnectionRepository.getOne(userId)

        session.setAttribute(USER_CONNECTION, connection)

        return connection
    }

    /**
     * Compile the best display name from the connection and the profile
     *
     * @param connection
     * @param profile
     * @return
     */
    internal fun getDisplayName(connection: UserConnection?, profile: UserProfile?): String {

        // The name is set differently in different providers so we better look in both places...
        return connection!!.displayname ?: profile!!.name
    }


    companion object {

       private const val USER_CONNECTION = "MY_USER_CONNECTION"
        private const val USER_PROFILE = "MY_USER_PROFILE"
    }

}

